chisel/README.md

```diff
@@ -0,0 +1,170 @@
+    - [x] Cloning
+- [ ] Optimizations.
+  - [ ] Speed up REPL execution time.
+- [x] Finish README.
+- [ ] First review.
+  - [ ] *add requested changes here*
+
+## Usage
+
+### Installation
+
+`chisel` is installed alongside Foundry cli commands!
+
+Simply run `foundryup` to install `chisel`!
+
+If you do not have `foundryup` installed, reference the Foundry [installation guide](../README.md#installation).
+
+### Cache Session
+
+While chisel sessions are not persistent by default, they can be saved to the cache via the builtin `flush` command from within the REPL.
+
+```bash
+$ chisel
+➜ uint a = 1;
+➜ uint b = a << 0x08;
+➜ !flush
+Saved session to cache with ID = 0.
+```
+
+### Loading a Previous Session
+
+Chisel allows you to load a previous session from your history.
+
+To view your history, you can run `chisel list` or `!list`. This will print a list of your previous sessions, identifiable by their index.
+
+You can also run `chisel view <id>` or `!view <id>` to view the contents of a specific session.
+
+To load a session, run `chisel load <id>` or use the `!load <id>` where `<id>` is a valid session index (eg 2 in the example below).
+
+```bash
+$ chisel history
+1. 2022-05-06 15:04:32 - chisel-0.json
+2. 2022-02-23 07:43:12 - chisel-1.json
+$ chisel view 2
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.17;
+
+contract REPL {
+    event KeccakEvent(bytes32 hash);
+
+    function run() public {
+      emit KeccakEvent(keccak256(abi.encode("Hello, world!")));
+    }
+}
+$ chisel load 2
+➜ ...
+```
+
+### Clearing the Cache
+
+To clear Chisel's cache (stored in `~/.foundry/cache/chisel`), use the `chisel clearcache` or `!clearcache` commands.
+
+```bash
+➜ !clearcache
+Cleared chisel cache!
+```
+
+### Toggling Traces
+
+By default, traces will only be shown if an input causes the call to the REPL contract to revert. To turn traces on
+regardless of the call result, use the `!traces` command or pass in a verbosity option of any level (`-v<vvvv>`) to
+the chisel binary.
+
+```bash
+➜ uint a
+➜ contract Test {
+    function get() external view returns (uint) {
+       return 256;
+    }
+}
+➜ Test t = new Test()
+➜ !traces
+Successfully enabled traces!
+➜ a = t.get()
+Traces:
+  [69808] 0xBd770416a3345F91E4B34576cb804a576fa48EB1::run()
+    ├─ [36687] → new <Unknown>@0xf4D9599aFd90B5038b18e3B551Bc21a97ed21c37
+    │   └─ ← 183 bytes of code
+    ├─ [315] 0xf4D9599aFd90B5038b18e3B551Bc21a97ed21c37::get() [staticcall]
+    │   └─ ← 0x0000000000000000000000000000000000000000000000000000000000000100
+    └─ ← ()
+
+➜ a
+Type: uint
+├ Hex: 0x100
+└ Decimal: 256
+```
+
+### Forking a Network
+
+To fork a network within your chisel session, use the `!fork <rpc-url>` command or supply a `--fork-url <rpc-url>` flag
+to the chisel binary. The `!fork` command also accepts aliases from the `[rpc_endpoints]` section of your `foundry.toml`,
+if chisel was launched in the root of a foundry project (ex. `!fork mainnet`).
```

Comment:
Does it also support env vars, e.g. `!fork MAINNET_RPC_URL`?

Reply:
Interpolation would be great. Maybe !fork ${MAINNET_RPC_URL}? We have a `fn interpolate` in the config code which should let you do that.

Reply:
Added using [RpcEndpoint::resolve](https://github.com/foundry-rs/foundry/blob/master/config/src/endpoints.rs#L87-L92) :+1:

Reply:
We should check that the RPC URL is valid at the time of fork and better handle the error. Using `!fork {ETH_RPC_URL}` works as expected, but if I forget the braces here's what happens: it panics and exists chisel once I try using it

<img width="780" alt="image" src="https://user-images.githubusercontent.com/17163988/198700725-e76ab8f9-85be-418b-b535-4cb980c8c2f6.png">